packages/@aws-cdk/core/lib/deployment-configuration.ts

```diff
@@ -0,0 +1,635 @@
+}
+
+/**
+ * Uses conventionally named roles and reify asset storage locations
+ *
+ * This DeploymentConfiguration is the only DeploymentConfiguration that generates
+ * an asset manifest, and is required to deploy CDK applications using the
+ * `@aws-cdk/app-delivery` CI/CD library.
+ *
+ * Requires the environment to have been bootstrapped with Bootstrap Stack V2.
+ */
+export class DefaultDeploymentConfiguration implements IDeploymentConfiguration {
+  private stack!: Stack;
+  private bucketName!: string;
+  private repositoryName!: string;
+  private deployActionRoleArn!: string;
+  private cloudFormationExecutionRoleArn!: string;
+  private assetPublishingRoleArn!: string;
+
+  private readonly assets: asset_schema.ManifestFile = {
+    version: asset_schema.AssetManifestSchema.currentVersion(),
+    files: {},
+    dockerImages: {},
+  };
+
+  constructor(private readonly props: DefaultDeploymentConfigurationProps = {}) {
+  }
+
+  public bind(stack: Stack): void {
+    this.stack = stack;
+
+    const qualifier = this.props.qualifier ?? 'hnb659fds';
+
+    // Function to replace placeholders in the input string as much as possible
+    //
+    // We replace:
+    // - ${Qualifier}: always
+    // - ${AWS::AccountId}, ${AWS::Region}: only if we have the actual values available
+    // - ${AWS::Partition}: never, since we never have the actual partition value.
+    const specialize = (s: string) => {
+      s = replaceAll(s, '${Qualifier}', qualifier);
+      return cxapi.EnvironmentPlaceholders.replace(s, {
+        region: resolvedOr(this.stack.region, cxapi.EnvironmentPlaceholders.CURRENT_REGION),
+        accountId: resolvedOr(this.stack.account, cxapi.EnvironmentPlaceholders.CURRENT_ACCOUNT),
+        partition: cxapi.EnvironmentPlaceholders.CURRENT_PARTITION,
+      });
+    };
+
+    // tslint:disable:max-line-length
+    this.bucketName = specialize(this.props.stagingBucketName ?? 'cdk-bootstrap-${Qualifier}-assets-${AWS::AccountId}-${AWS::Region}');
+    this.repositoryName = specialize(this.props.ecrRepositoryName ?? 'cdk-bootstrap-${Qualifier}-container-assets-${AWS::AccountId}-${AWS::Region}');
+    this.deployActionRoleArn = specialize(this.props.deployActionRoleArn ?? 'arn:${AWS::Partition}:iam::${AWS::AccountId}:role/cdk-bootstrap-deploy-action-role-${AWS::AccountId}-${AWS::Region}');
+    this.cloudFormationExecutionRoleArn = specialize(this.props.cloudFormationExecutionRole ?? 'arn:${AWS::Partition}:iam::${AWS::AccountId}:role/cdk-bootstrap-cfn-exec-role-${AWS::AccountId}-${AWS::Region}');
+    this.assetPublishingRoleArn = specialize(this.props.assetPublishingRoleArn ?? 'arn:${AWS::Partition}:iam::${AWS::AccountId}:role/cdk-bootstrap-publishing-role-${AWS::AccountId}-${AWS::Region}');
+    // tslint:enable:max-line-length
+  }
+
+  public addFileAsset(asset: FileAssetSource): FileAssetLocation {
+    const objectKey = asset.sourceHash + '.zip';
+
+    // Add to manifest
+    this.assets.files![asset.sourceHash] = {
```

Comment:
I am really uneasy about these `!` everywhere here. I have been bit every time I used these. Maybe instead of holding the entire manifest as a property, you can just hold a map of `files` and `images` as private properties and synthesize the manifest only later?